src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+                )?));
+            }
+            Ok(Arc::new(
+                decoded_results
+                    .iter()
+                    .collect::<arrow::array::BooleanArray>(),
+            ) as ArrayRef)
+        }
+
+        CreateFunctionDataType::F64 | CreateFunctionDataType::DOUBLE => {
+            decode_udf_result_primitive_array::<arrow::datatypes::Float64Type>(
+                encoded_results,
+                &|v| {
+                    v.as_f64().ok_or(DataFusionError::Internal(format!(
+                        "Expected to find f64 value, but received {:?} instead",
+                        v
+                    )))
+                },
+            )
+            .map(|a| Arc::new(a) as ArrayRef)
+        }
+        CreateFunctionDataType::F32
+        | CreateFunctionDataType::REAL
+        | CreateFunctionDataType::FLOAT => decode_udf_result_primitive_array::<
+            arrow::datatypes::Float32Type,
+        >(encoded_results, &|v| match v {
+            Value::F32(n) => Ok(*n),
+            _ => Err(DataFusionError::Internal(format!(
+                "Expected to find f64 value, but received {:?} instead",
+                v
+            ))),
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::DECIMAL {
+            precision: _,
+            scale: _,
+        } => {
+            let mut decoded_results = Vec::with_capacity(encoded_results.len());
+            for i in encoded_results {
+                let s = i.as_str().ok_or(DataFusionError::Internal(format!(
+                    "Expected to find str value, received {:?} instead",
+                    &i
+                )))?;
+                decoded_results.push(Some(s.parse::<i128>().map_err(|e| {
+                    DataFusionError::Internal(format!(
+                        "Error parsing string to i128: {:?}",
+                        e
+                    ))
+                })?));
+            }
+            Ok(Arc::new(
+                decoded_results
+                    .iter()
+                    .collect::<arrow::array::Decimal128Array>(),
+            ) as ArrayRef)
+        }
+    }
+}
+
+fn make_scalar_function_wasm_messagepack(
+    module_bytes: &[u8],
+    function_name: &str,
+    input_types: Vec<CreateFunctionDataType>,
+    return_type: CreateFunctionDataType,
+) -> Result<ScalarFunctionImplementation> {
+    // Similar to make_scalar_function_from_wasm, this function should verify
+    // that the module can be loaded and the UDF export is found before
+    // returning a Result.
+    let function_name = function_name.to_owned();
+    let module_bytes = module_bytes.to_owned();
+    let _outer_instance = WasmMessagePackUDFInstance::new(&module_bytes, &function_name)
+        .map_err(|err| {
+            DataFusionError::Internal(format!(
+                "Error initializing WASM + MessagePack UDF {:?}: {:?}",
+                function_name, err
+            ))
+        })?;
```

Comment:
This variable seems to be unused?

Reply:
That's the part where the module's exports are verified before constructing `inner` to fail early if there's something wrong with the wasm module (eg: no export named `concat2`).